tensorboard/plugins/debugger_v2/debugger_v2_plugin.py

```diff
@@ -110,3 +116,72 @@ def serve_execution_digests(self, request):
             return http_util.Respond(
                 request, {"error": str(e)}, "application/json", code=400,
             )
+
+    @wrappers.Request.application
+    def serve_source_files_list(self, request):
+        """Serves a list of all source files involved in the debugged program."""
+        experiment = plugin_util.experiment_id(request.environ)
+        run = request.args.get("run")
+        if run is None:
+            return _missing_run_error_response(request)
+        run_tag_filter = debug_data_provider.source_file_list_run_tag_filter(
+            run
+        )
+        blob_sequences = self._data_provider.read_blob_sequences(
+            experiment, self.plugin_name, run_tag_filter=run_tag_filter
+        )
+        tag = next(iter(run_tag_filter.tags))
+        return http_util.Respond(
+            request,
+            self._data_provider.read_blob(blob_sequences[run][tag][0].blob_key),
+            "application/json",
+        )
+
+    @wrappers.Request.application
+    def serve_source_file(self, request):
+        """Serves the content of a given source file.
+
+        The source file is referred to by the index in the list of all source files
+        invovled in the execution of the debugged program, which is available via the
+        `serve_source_files_list()`  serving route.
+
+        Args:
+          request: HTTP request.
+
+        Returns:
+          Response to the request.
+        """
+        experiment = plugin_util.experiment_id(request.environ)
+        run = request.args.get("run")
+        if run is None:
+            return _missing_run_error_response(request)
+        index = request.args.get("index")
+        # TOOD(cais): When the need arises, support serving a subset of a
+        # source file's lines.
+        if index is None:
+            return http_util.Respond(
+                request,
+                {"error": "index is not provided for source file content"},
+                "application/json",
+                code=400,
+            )
+        index = int(index)
+        run_tag_filter = debug_data_provider.source_file_run_tag_filter(
+            run, index
+        )
+        blob_sequences = self._data_provider.read_blob_sequences(
+            experiment, self.plugin_name, run_tag_filter=run_tag_filter
+        )
+        tag = next(iter(run_tag_filter.tags))
+        try:
+            return http_util.Respond(
+                request,
+                self._data_provider.read_blob(
+                    blob_sequences[run][tag][0].blob_key
+                ),
+                "application/json",
+            )
+        except IndexError as e:
+            return http_util.Respond(
+                request, {"error": str(e)}, "application/json", code=400,
```

Comment:
This is a bit of a confusing message—the “list index out of range”
doesn’t mean that the user-provided index was out of bounds but rather
that the `[0]` on line 180 above failed, right? The raw Python error
message is an implementation detail; do you think that it would be more
appropriate to raise “no source file %d for run %s” as a 404?